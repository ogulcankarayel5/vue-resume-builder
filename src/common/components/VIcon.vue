<template>
  <component :is="iconComponent" :color="color" role="img" />
</template>

<script lang="ts">
import Vue, { defineComponent } from "vue";

const icons: any = {
  reveal: require("@/common/icons/RevealIcon.vue").default,
  revealHide: require("@/common/icons/RevealHideIcon.vue").default,
  resume: require("@/common/icons/ResumeIcon.vue").default,
  tick: require("@/common/icons/TickIcon.vue").default,
  close: require("@/common/icons/CloseIcon.vue").default,
  google: require("@/common/icons/GoogleIcon.vue").default,
  nav: require("@/common/icons/NavIcon.vue").default,
};

export default defineComponent({
  name: "VIcon",
  props: {
    name: {
      type: String,
      required: true,
    },
    color: {
      type: String,
      required: false,
    },
  },
  computed: {
    iconComponent() {
      return icons[this.name as any];
    },
  },
});
</script>
