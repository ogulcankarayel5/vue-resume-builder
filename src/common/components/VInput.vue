<template>
  <div class="w-80 flex flex-col" :class="containerClassName">
    <label data-test-id="label" v-if="label">{{ label }}</label>
    <div
      class="relative flex items-center"
      :class="{ 'mt-3': label !== undefined }"
    >
      <input
        class="flex-1 p-3 shadow-2xl rounded-lg focus:outline-none"
        :class="className"
        v-bind="$attrs"
        :value="value"
        @input="update"
        :type="showPassword ? 'text' : type"
        data-test-id="input"
      />
      <div
        v-if="showHideIcon"
        class="absolute top-2/4 transform -translate-y-2/4 right-2 cursor-pointer"
        @click="togglePassword"
        data-test-id="password"
      >
        <v-icon data-test-id="reveal" v-if="showPassword" name="reveal" />
        <v-icon data-test-id="revealHide" v-else name="revealHide" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
export default defineComponent({
  inheritAttrs: false,
  name: "VInput",
  data() {
    return {
      showPassword: false,
    };
  },
  props: {
    value: {
      required: false,
    },
    type: {
      type: String,
      default: "text",
    },
    className: {
      type: [String, Array, Object],
      required: false,
    },
    containerClassName: {
      type: [String, Array, Object],
      required: false,
    },
    showHideIcon: {
      type: Boolean,
      default: false,
    },
    label: {
      type: String,
      required: false,
    },
  },

  model: {
    prop: "value",
    event: "update",
  },
  methods: {
    update(event: Event): void {
      this.$emit("update", (event.target as HTMLInputElement).value);
    },
    togglePassword(): void {
      this.showPassword = !this.showPassword;
    },
  },
});
</script>

<style></style>
