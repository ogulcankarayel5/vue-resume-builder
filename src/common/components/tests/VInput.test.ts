import VInput from "@/common/components/VInput.vue";
import { shallowMount } from "@vue/test-utils";
import VIcon from "@/common/components/VIcon.vue";

const factory = (values = {}, other = {}) => {
  return shallowMount(VInput, {
    data() {
      return {
        ...values,
      };
    },
    ...other,
    components: {
      VIcon,
    },
  });
};

describe("VInput", () => {
  it("snapshot", () => {
    const wrapper = factory();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render password icon if showHideIcon is true", () => {
    const wrapper = factory({}, { propsData: { showHideIcon: true } });
    expect(wrapper.find('[data-test-id="password"]').exists()).toBe(true);
  });

  it("should change icon according to showPassword state", async () => {
    const wrapper = factory(
      { showPassword: false },
      { propsData: { showHideIcon: true } }
    );

    expect(wrapper.find('[data-test-id="reveal"]').exists()).toBe(false);
    expect(wrapper.find('[data-test-id="revealHide"]').exists()).toBe(true);

    await wrapper.find('[data-test-id="password"]').trigger("click");

    expect(wrapper.find('[data-test-id="reveal"]').exists()).toBe(true);
    expect(wrapper.find('[data-test-id="revealHide"]').exists()).toBe(false);
  });

  it("should render label correctly",()=>{
    const wrapper = factory({},{ propsData: { label: "test" } });
    expect(wrapper.find('[data-test-id="label"]').exists()).toBe(true);
  });

  it("input type should be text when showPassword state is true", async () => {
    const wrapper = factory(
      { showPassword: false },
      { propsData: { showHideIcon: true } }
    );


    await wrapper.find('[data-test-id="password"]').trigger("click");

    expect(wrapper.find('[data-test-id="input"]').attributes("type")).toBe("text");
  });
});
