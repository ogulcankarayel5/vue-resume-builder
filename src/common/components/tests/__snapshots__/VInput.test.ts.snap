// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VInput snapshot 1`] = `
<div
  class="w-80 flex flex-col"
>
  <!---->
   
  <div
    class="relative flex items-center"
  >
    <input
      class="flex-1 p-3 shadow-2xl rounded-lg focus:outline-none"
      data-test-id="input"
      type="text"
    />
     
    <!---->
  </div>
</div>
`;
